Sources/Commands/SwiftPackageTool.swift

```diff
@@ -863,6 +865,138 @@ extension SwiftPackageTool {
+
+            // The `plugins` directory is inside the workspace's main data directory, and contains all temporary files related to all plugins in the workspace.
+            let dataDir = try swiftTool.getActiveWorkspace().location.workingDirectory
+            let pluginsDir = dataDir.appending(component: "plugins")
+
+            // The `cache` directory is in the plugins directory and is where the plugin script runner caches compiled plugin binaries and any other derived information.
+            let cacheDir = pluginsDir.appending(component: "cache")
+            let pluginScriptRunner = DefaultPluginScriptRunner(cacheDir: cacheDir, toolchain: try swiftTool.getToolchain().configuration, enableSandbox: false)
+
+            // The `outputs` directory contains subdirectories for each combination of package, target, and plugin. Each usage of a plugin has an output directory that is writable by the plugin, where it can write additional files, and to which it can configure tools to write their outputs, etc.
+            let outputDir = pluginsDir.appending(component: "outputs")
+
+            // Build the map of tools that are available to the plugin. This should include the tools in the executables in the toolchain, as well as any executables the plugin depends on (built executables as well as prebuilt binaries).
+            // FIXME: At the moment we just pass the built products directory for the host. We will need to extend this with a map of the names of tools available to each plugin. In particular this would not work with any binary targets.
+            let builtToolsDir = dataDir.appending(components: "plugin-tools")
+
+            // Create the cache directory, if needed.
+            try localFileSystem.createDirectory(cacheDir, recursive: true)
+            
+            // FIXME: Need to determine the correct root package.
+            
+            // FIXME: Need to
+            // Determine the tools to which this plugin has access, and create a name-to-path mapping from tool
+            // names to the corresponding paths. Built tools are assumed to be in the build tools directory.
+            let accessibleTools = plugin.accessibleTools(for: pluginScriptRunner.hostTriple)
+            let toolNamesToPaths = accessibleTools.reduce(into: [String: AbsolutePath](), { dict, tool in
+                switch tool {
+                case .builtTool(let name, let path):
+                    dict[name] = builtToolsDir.appending(path)
+                case .vendedTool(let name, let path):
+                    dict[name] = path
+                }
+            })
+
+            // Run the plugin.
+            let result = try plugin.invoke(
+                action: .performCommand(targets: Array(targets.values), arguments: arguments),
+                package: packageGraph.rootPackages[0], // FIXME: This should be the package that contains all the targets (and we should make sure all are in one)
```

Comment:
I believe SwiftTool has several examples or even a func to return the root package

Reply:
I thought so too, but saw a couple of other cases of `rootPackages[0]` so followed that.  I'll take a look.

Reply:
Actually I see `graph.rootPackages[0]` used in most of the commands.  Do you think it would be worth adding a cover for this?